### PR TITLE
GDScript: Add raw string literals (r-strings)

### LINKED ISSUE #5362: Add raw strings literals to GDScript
### Describe the project you are working on

Various projects.

### Describe the problem or limitation you are having in your project

When working with regular expressions containing backslashes, you need to escape each backslash, and the readability of the regular expression is noticeably reduced.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add raw strings literals, where only the opening quote character **must** be escaped. You **can** escape the backslash, but you **don't have to** if it's followed by a character that doesn't match the opening quote. _EDIT: Backslashes and quotes cannot be really escaped in Python r-strings, since they stay in the result._ Also, escape sequences like `\n` and `\t` do not work in r-strings (they remain unchanged). This is how r-strings work in Python, and there is a [proposal for f-strings](https://github.com/godotengine/godot-proposals/issues/157).

On the other hand, if you want `\\`, then even in r-strings you must use `\\\\` (r-strings only get rid of doubling a single `\`, like `\d`). _EDIT: I was wrong, see above._ Perhaps some other option would be better, such as doubling the quote to escape (as in SQL and Pascal). Or regular expression literals like in JavaScript. _EDIT: Given that string literals are already Python-like in GDScript (triple quotes for multiline strings), r- and f-strings probably should be the same as in Python._

---

**LAST EDIT:** Here are the "escaping" rules in r-strings:

```
\" -> \" (or \' -> \' if in r'...')
\\ -> \\ (this allows you \\\")
\ -> \
```
Unlike regular strings, r-strings converts special sequences to themselves.

---

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```
'\\d+(?:\\.\\d+)?|\\+|-|\\*\\*?|/|\\(|\\)|(.)'
r'\d+(?:\.\d+)?|\+|-|\*\*?|/|\(|\)|(.)'

Python r-strings:
print(r'" \' \ \\') # " \' \ \\
print(r"\" ' \ \\") # \" ' \ \\

You cannot represent a string containing an even number of backslashes in a row before a quote.
```

Also multiline string literals (`r"""string"""` and `r'''string'''`).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is about GDScript syntax.

### Is there a reason why this should be core and not an add-on in the asset library?

This cannot be done with an add-on.

### PR DESCRIPTION
**Pull request by dalexeev**
* Add support for r-strings to the tokenizer.
  * There are actually few changes, just added 1 indent level (use "Hide whitespace").
  * The behavior is the same as in Python (it is not well described in the proposal).
* Add syntax highlighting.
  * Known issue: After a line break, escape sequences are highlighted again (like in regular strings).
  * Bonus: Fixed highlighting of Unicode sequences in regular strings.

Closes godotengine/godot-proposals#5362.

### CODE REVIEW

**Comment by MewPurPur** [comment]
Code looks good to me. Implementation idk, I don't get if it's possible to have a raw string with both

> Known issue: After a line break, escape sequences are highlighted again (like in regular strings).

I don't think there's a neat way to solve this without making the highlighter remember things between lines. It might be solved by an extra region paintover, aka getting the string delimiters, prepending them all with an "r", and getting all slices starting with `r + str_delim` and ending with `str_delim` (not sure how to do this last part, I don't think we have a version of `get_slice()` with different starting and ending delimiters). It might be an overall simplification actually, since it removes the need for `in_raw_string` in `get_line_syntax_highlighting`.

**Comment by dalexeev** [comment]
> Code looks good to me. Implementation idk, I don't get if it's possible to have a raw string with both " and ' in it. Though I guess that's inherent to the concept.

Yes, you can combine quotes of different types.



> I don't think there's a neat way to solve this without making the highlighter remember things between lines.

I think we can think about this later, the highlighter has other issues related to newlines and not only. I think the main use case for r-strings is regular expressions.

**Review by MewPurPur — COMMENTED** [review]
Everything looks good, but what's up with having a bunch of navigation-related files in the changed files? Did you add them by accident?

**Comment by MewPurPur** [comment]
Oh I'm so sorry, I didn't realize I was looking at something else  ;w;

**Comment by dalexeev** [comment]
**How it works?**

In regular string literals, you can use special characters (`\n`, `\t`) and escape quotes, special characters and the escape character itself (backslash). Thanks to this, you get a cleaner literal (special characters instead of whitespaces, zero-width characters, etc.) and the ability to represent _any_ string

r-string literals have the advantage that you get the same string that you see in the source code between `r"` and `"` (or between `r"""` and `"""`, same with single quotes). But you lose the ability to represent _any_ string (some strings become unrepresentable, but they are almost never needed).

For example, in the literal `r"..."` you cannot use a single `"`, since it is the terminator of the string literal. But if it is preceded by a backslash, then it is allowed (`r"\""`). In other words, "escaping" in r-strings exists, but the sequences are escaped into themselves. (Note that you can almost always use other quotes or triple quotes, although in general you still cannot represent an arbitrary string.)

There is another type of raw string literals. R-strings in C++ or Nowdoc syntax in PHP. You can specify a sequence of characters that does not appear in a string, and terminate the literal with the same sequence (in PHP) or the reverse sequence (C++).

```c++
   R"(")"
//  <<^>>

   R"*()")*"
//  <<<^^>>>

   R"**()*")**"
//  <<<<^^^>>>>
```

But I don't think we need a solution as universal as C++ or PHP. String literals in GDScript are already very similar to Python, perhaps we should just copy them like we did in this PR?

**Comment by adamscott** [comment]
> There is another type of raw string literals. R-strings in C++ or Nowdoc syntax in PHP. You can specify a sequence of characters that does not appear in a string, and terminate the literal with the same sequence (in PHP) or the reverse sequence (C++).
> 
> ```c++
>    R"(")"
> //  <<^>>
> 
>    R"*()")*"
> //  <<<^^>>>
> 
>    R"**()*")**"
> //  <<<<^^^>>>>
> ```
> 
> But I don't think we need a solution as universal as C++ or PHP. String literals in GDScript are already very similar to Python, perhaps we should just copy them like we did in this PR?

Why not both? Lowercase "r" would be the current implementation, and uppercase "R" would be the C++/PHP one?

**Comment by dalexeev** [comment]
> Why not both?

This is very rarely required, in my opinion. It is a more complex syntax, while GDScript is a simple language. I think that a language like GDScript should not have many different syntaxes for similar purposes, it can confuse users.

**Review by adamscott — APPROVED** [review]
Didn't check the code yet, but the GDScript team approve the PR.

File: modules/gdscript/gdscript_tokenizer.cpp [hunk 1]
```diff
@@ -905,144 +914,164 @@ GDScriptTokenizer::Token GDScriptTokenizer::string() {
...
+				}
+				// Parse UTF-16 pair.
+				if (valid_escape) {
+					if ((escaped & 0xfffffc00) == 0xd800) {
+						if (prev == 0) {
+							prev = escaped;
+							prev_pos = column - 2;
+							continue;
+						} else {
+							Token error = make_error("Invalid UTF-16 sequence in string, unpaired lead surrogate");
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_tokenizer.cpp:1044]:
```suggestion
							Token error = make_error("Invalid UTF-16 sequence in string, unpaired lead surrogate.");
```

**Review by adamscott — APPROVED** [review]
Approved by the GDScript team, r-strings as presented, are a good addition to the language.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**